barman/cli.py

```diff
@@ -922,21 +922,34 @@ def recover(args):
                 output.error("Cannot parse recovery staging path: %s", str(exc))
                 output.close_and_exit()
             server.config.recovery_staging_path = recovery_staging_path
-        # If the backup is compressed but there is no recovery_staging_path
-        # then this is an error - the user *must* tell barman where recovery
-        # data can be staged.
+
+        # If no staging path is found then we need to display a proper error
+        # according to the backup type (compressed or incremental)
         if server.config.recovery_staging_path is None:
```

Comment:
IIRC `recovery_staging_path` is a path in the target host when used for compressed backups.
So, for compressed backups, it's an intermediate directory where Barman copies the compressed backups to, then later the backup is decompressed and moved to the final destination pointed by `barman recover`.

A default value could make the experience when recovering incremental backup more seamless. But I don't think we should assume paths inside Barman home for that GUC -- unless we set a "default" only when recovering an incremental backup. I mean something like:

* Is incremental backup recovery?
  * Is recovery staging path not set?
    * Maybe issue a warning
    * Set recovery staging path as you proposed

Reply:
I think @gustabowill wanted to follow the tarball standards, so it does not have a default value, but requires a recovery_staging_path, right?? IMO setting the recovery_staging_path for the recovery is a little useless in this case because we combine in the staging and delete it after the copying/moving is done to the final destination. So for the UX, it is almost hidden (only have messages as output.info) because he/she won't have access to the combined backup directory after all. 
But i think a default value is nice to have! Then we do not require the user to set recovery_staging_path at all, only if he/she actually needs it. Then we could follow @barthisrael previous suggestion.

What are your thoughts? Can i go ahead with @barthisrael suggestion? @martinmarques @barthisrael  @gcalacoci  @gustabowill  @bleidens

Reply:
my question was around "should we change the actual behavior?"
the fact is: if we don't set it up, we error out. 

Is this what we want or do we want to change?
> IMO setting the recovery_staging_path for the recovery is a little useless in this case because we combine in the staging and delete it after the copying/moving is done to the final destination.

True that is something that get's whiped out after the recovery.
BUT, while you are recovering in an emergency, the less stuff you configure, the better.

BTW, good feedback so far guys.

Reply:
> True that is something that get's whiped out after the recovery.
> BUT, while you are recovering in an emergency, the less stuff you configure, the better.

Right!

That's why I thought of having a default only when recovering an incremental backup, which default can be more predictable/acceptable in that case.

A `WARNING`, or even an `INFO` message, would be cool to have in the output, just so the user is aware of what's going on.
I think it is worth it a note in the docs later, too, so the user is aware that we will temporarily consume some considerable disk space when combining backups, and the location is set by `--recovery-staging-path`, defaulting to location "X".

Reply:
I'm starting to wonder if it our previous idea of having a specific config for the combine staging directory was maybe the best option 🤔. It seems good to have a default value for this, but it seems like it makes sense for one use of `--recovery-staging-path` and not all, maybe having a specific param for the combine directory would make it easier to set a default and also facilitate explanations in the docs later?

Reply:
I strongly feel that having a default value will lead to users not thinking about it, and having it fill the disk where they have their backups and WALs.
I vote to make this a mandatory option when recovering

Reply:
well, before that was something used only for tar stuff... that was sort of ok... but the way I see incremental going in the future is possibly going to be how people are going to do backups by default, because is faster and uses less space.
THIS changes the importance of the option, IMHO.

@barthisrael having a default only for incrementals is possibly confusing IMHO.

@martinmarques I see your point, and is something that "scares me" too... that's why I started this conversation.

Reply:
> well, before that was something used only for tar stuff... that was sort of ok... but the way I see incremental going in the future is possibly going to be how people are going to do backups by default, because is faster and uses less space.
THIS changes the importance of the option, IMHO.
> 
> @barthisrael having a default only for incrementals is possibly confusing IMHO.

Yeah, maybe if we decide for using a default, a separate option, let's say `--combine-staging-path` would suit best now. As I mentioned initially, I thought there would be room for reusing `--recovery-staging-path`, until we eventually find that would not fit 😆 

> @martinmarques I see your point, and is something that "scares me" too... that's why I started this conversation.
Yeah, when reviewing the PR I was ok with forcing the user to choose a path as that can consume a lot of disk space. I'm afraid that if we use defaults, that can be "destructive" for the user environment.

Reply:
I propose to continue with the actual behaviour and then test the whole incremental backup and recovery experience manually a couple of times to get a feeling of the real-world usage.

After that, we can re-evaluate this. I don't want to introduce too much complexity or problems now

@martinmarques @andremagui @gustabowill @bleidens @barthisrael

Reply:
I think that as long as you can put a value as a configuration option in the .conf, the user will only get an exception on the first try without setting the staging dir, then they configure it and done.